### PR TITLE
Generate different types of fsms.

### LINKED ISSUE #60: Generate efficient FSMs for slow modules
When a module does not have an II=1, we should be able to generate smaller, counter-based FSMs. This is particularly important if we're interacting with modules of high-latency since we may need to a long time and the generated pipelined FSM can get quite large.

### PR DESCRIPTION
**Pull request by UnsignedByte**
Resolves #60.

Separates FSM generation to another file, and adds more efficient fsms designs.

### CODE REVIEW

File: src/ir_passes/lower/build_ctx.rs [hunk 1]
```diff
@@ -227,117 +213,25 @@ impl<'a> BuildCtx<'a> {
     /// Attempts to declare an fsm component (if not already declared) in the [Binding] stored by this [BuildCtx]
     /// and creates an [Fsm] from this [calyx::Component] FSM and stores it in the [BuildCtx]
     pub fn insert_fsm(&mut self, event: ir::EventIdx, states: u64) {
+        let evt = self.comp.get(event);
         // Construct an fsm iff the event is connected to an interface port
-        if self.comp.get(event).has_interface {
-            self.declare_fsm(states);
+        if evt.has_interface {
+            let ir::TimeSub::Unit(delay) = evt.delay else {
+                unreachable!("Non-unit delays should have been compiled away.");
```
Comment (rachitnigam) [inline src/ir_passes/lower/build_ctx.rs:220]: Let's use `comp.internal_error` instead of an explicit `panic!`

File: src/ir_passes/lower/fsm.rs [hunk 2]
```diff
@@ -1,17 +1,406 @@
...
+        self.add_opt(states, delay, Some(lib))
+    }
+
+    fn add_opt(
+        &'l mut self,
+        states: u64,
+        delay: u64,
+        lib: Option<&'l calyx::LibrarySignatures>,
+    ) -> &'l calyx::Component {
+        // TODO: Find a better metric to decide which type of fsm to generate.
```
Comment (rachitnigam) [inline src/ir_passes/lower/fsm.rs:45]: Let's create a tracking issue for this. We want to come up with some simple threshold to see when we can generate simple, shift-register-based FSMs vs the more complex counter ones. For extremely large delays, where the module is essentially sequential, we definitely want counter-based. On the other end of the spectrum, almost completely pipelined modules should probably use shift register based ones. We'll need to do some measurements to figure out which one to choose.

File: src/ir_passes/lower/fsm.rs [hunk 3]
```diff
@@ -1,17 +1,406 @@
...
+                );
+
+                drop(this);
+                comp
+            })
+    }
+
+    /// Helper function that generates a [calyx:Component] for a counter with `n` states.
+    fn add_counter(
+        &'l mut self,
```
Comment (rachitnigam) [inline src/ir_passes/lower/fsm.rs:181]: Is the explicit lifetime needed here?

File: src/ir_passes/lower/fsm.rs [hunk 4]
```diff
@@ -1,17 +1,406 @@
...
+            // gets the number of bits needed to represent the counter state.
+            let bitwidth = (64 - (states - 1).leading_zeros()) as u64;
+
+            let mut comp = FsmBind::fsm_comp("counter", states);
+            let mut builder =
+                calyx::Builder::new(&mut comp, lib.unwrap()).not_generated();
+
+            let state = builder.add_primitive("state", "std_reg", &[bitwidth]);
+
+            let add = builder.add_primitive("add", "std_add", &[bitwidth]);
```
Comment (rachitnigam) [inline src/ir_passes/lower/fsm.rs:195]: Consider using [`calyx_ir`'s macros](https://docs.rs/calyx-ir/0.4.0/calyx_ir/#macros) here to add new primitives and build complex assignments. We've always pondered how to we can create better ways of writing down complex guards using a macro so maybe we should create an issue in the Calyx repo for that as well.

File: src/ir_passes/lower/fsm.rs [hunk 5]
```diff
@@ -1,17 +1,406 @@
...
+
+            // Constant signal
+            structure!(builder;
+                let signal_on = constant(1, 1);
+                let one = constant(1, bitwidth);
+                let zero = constant(0, bitwidth);
+            );
```
Comment (rachitnigam) [inline src/ir_passes/lower/fsm.rs:202]: Oops, I see you already use them! In that case, the above should be folded into this `structure!` call.

File: src/ir_passes/lower/fsm.rs [hunk 6]
```diff
@@ -1,17 +1,406 @@
...
+
+            let done = builder.add_primitive("done", "std_reg", &[1]);
+
+            // This component's interface
+            let this = builder.component.signature.borrow();
+
+            // checks if the counter is currently on the final state.
+            // state == _{n-1}
+            let rst_check: Guard<Nothing> =
+                Guard::Port(state.borrow().get("out")).eq(Guard::Port(
```
Comment (rachitnigam) [inline src/ir_passes/lower/fsm.rs:219]: The `guard!` macro can be used here.

File: src/ir_passes/lower/fsm.rs [hunk 7]
```diff
@@ -1,17 +1,406 @@
...
+            let go_check: Guard<Nothing> = Guard::Port(this.get("go")).or(
+                Guard::Port(state.borrow().get("out"))
+                    .eq(Guard::Port(zero.borrow().get("out")))
+                    .not(),
+            );
+
+            // add base assignments
+            builder.component.continuous_assignments.extend([
+                // build assignments for state+1
+                builder.build_assignment(
```
Comment (rachitnigam) [inline src/ir_passes/lower/fsm.rs:234]: These assignments should be implemented using `build_assignments!`

**Review by rachitnigam — APPROVED** [review]
LGTM! Couple of style comments. Resolve at your discretion and merge! One optimization that we should implement in the future (let's open an issue to discuss this). If you compile `tests/compile/deserialize.fil`, you'll notice we generate guards like this:
```
component comp1<"nointerface"=1>(@data p10: 32, @data p11: 32, @data p12: 32, @data p13: 32, @data p14: 32, @data p15: 32, @fil_event ev0: 1, @reset reset: 1, @clk clk: 1) -> (@data p16: 32, @data p17: 32, @data p18: 32, @data p19: 32, @data p20: 32, @data p21: 32) {
  cells {
    ev00 = fsm_14();
    @data inst0 = Register(32);
    @data inst1 = Register(32);
    @data inst2 = Register(32);
    @data inst3 = Register(32);
  }
  wires {
    ev00.go = ev0;
    inst0.write_en = ev00._0 ? 1'd1;
    inst1.write_en = ev00._0 ? 1'd1;
    inst2.write_en = ev00._3 ? 1'd1;
    inst3.write_en = ev00._3 ? 1'd1;
    inst0.in = ev00._0 | ev00._1 | ev00._2 | ev00._3 | ev00._4 | ev00._5 ? p10;
    p16 = ev00._1 | ev00._2 | ev00._3 | ev00._4 | ev00._5 | ev00._6 | ev00._7 | ev00._8 | ev00._9 | ev00._10 | ev00._11 | ev00._12 | ev00._13 ? inst0.out;
    inst1.in = ev00._0 | ev00._1 | ev00._2 | ev00._3 | ev00._4 | ev00._5 ? p11;
    p17 = ev00._1 | ev00._2 | ev00._3 | ev00._4 | ev00._5 | ev00._6 | ev00._7 | ev00._8 | ev00._9 | ev00._10 | ev00._11 | ev00._12 | ev00._13 ? inst1.out;
    inst2.in = ev00._3 | ev00._4 | ev00._5 | ev00._6 | ev00._7 | ev00._8 ? p12;
    p18 = ev00._4 | ev00._5 | ev00._6 | ev00._7 | ev00._8 | ev00._9 | ev00._10 | ev00._11 | ev00._12 | ev00._13 ? inst2.out;
    inst3.in = ev00._3 | ev00._4 | ev00._5 | ev00._6 | ev00._7 | ev00._8 ? p13;
    p19 = ev00._4 | ev00._5 | ev00._6 | ev00._7 | ev00._8 | ev00._9 | ev00._10 | ev00._11 | ev00._12 | ev00._13 ? inst3.out;
    p20 = ev00._6 | ev00._7 | ev00._8 | ev00._9 | ev00._10 | ev00._11 ? p14;
    p21 = ev00._6 | ev00._7 | ev00._8 | ev00._9 | ev00._10 | ev00._11 ? p15;
  }
  control {}
}
```

Essentially, we're having to check whether the FSM states are in `[0, 5]` or `[1, 13]`. It occurs to me that in a counter-based FSM, we could generate this comparison much more efficiently by exposing the internal `count` signal as well. In the future, we should expose this signal and generate simpler guards for these slow modules.

I've asked to create an issue about heuristics as well. Long-term, we want to start thinking about a low-level IR that can represent these optimization opportunities (or wether we can do this in the lowering itself.)

**Comment by UnsignedByte** [comment]
> LGTM! Couple of style comments. Resolve at your discretion and merge! One optimization that we should implement in the future (let's open an issue to discuss this). If you compile `tests/compile/deserialize.fil`, you'll notice we generate guards like this:
> 
> ```
> component comp1<"nointerface"=1>(@data p10: 32, @data p11: 32, @data p12: 32, @data p13: 32, @data p14: 32, @data p15: 32, @fil_event ev0: 1, @reset reset: 1, @clk clk: 1) -> (@data p16: 32, @data p17: 32, @data p18: 32, @data p19: 32, @data p20: 32, @data p21: 32) {
>   cells {
>     ev00 = fsm_14();
>     @data inst0 = Register(32);
>     @data inst1 = Register(32);
>     @data inst2 = Register(32);
>     @data inst3 = Register(32);
>   }
>   wires {
>     ev00.go = ev0;
>     inst0.write_en = ev00._0 ? 1'd1;
>     inst1.write_en = ev00._0 ? 1'd1;
>     inst2.write_en = ev00._3 ? 1'd1;
>     inst3.write_en = ev00._3 ? 1'd1;
>     inst0.in = ev00._0 | ev00._1 | ev00._2 | ev00._3 | ev00._4 | ev00._5 ? p10;
>     p16 = ev00._1 | ev00._2 | ev00._3 | ev00._4 | ev00._5 | ev00._6 | ev00._7 | ev00._8 | ev00._9 | ev00._10 | ev00._11 | ev00._12 | ev00._13 ? inst0.out;
>     inst1.in = ev00._0 | ev00._1 | ev00._2 | ev00._3 | ev00._4 | ev00._5 ? p11;
>     p17 = ev00._1 | ev00._2 | ev00._3 | ev00._4 | ev00._5 | ev00._6 | ev00._7 | ev00._8 | ev00._9 | ev00._10 | ev00._11 | ev00._12 | ev00._13 ? inst1.out;
>     inst2.in = ev00._3 | ev00._4 | ev00._5 | ev00._6 | ev00._7 | ev00._8 ? p12;
>     p18 = ev00._4 | ev00._5 | ev00._6 | ev00._7 | ev00._8 | ev00._9 | ev00._10 | ev00._11 | ev00._12 | ev00._13 ? inst2.out;
>     inst3.in = ev00._3 | ev00._4 | ev00._5 | ev00._6 | ev00._7 | ev00._8 ? p13;
>     p19 = ev00._4 | ev00._5 | ev00._6 | ev00._7 | ev00._8 | ev00._9 | ev00._10 | ev00._11 | ev00._12 | ev00._13 ? inst3.out;
>     p20 = ev00._6 | ev00._7 | ev00._8 | ev00._9 | ev00._10 | ev00._11 ? p14;
>     p21 = ev00._6 | ev00._7 | ev00._8 | ev00._9 | ev00._10 | ev00._11 ? p15;
>   }
>   control {}
> }
> ```
> 
> Essentially, we're having to check whether the FSM states are in `[0, 5]` or `[1, 13]`. It occurs to me that in a counter-based FSM, we could generate this comparison much more efficiently by exposing the internal `count` signal as well. In the future, we should expose this signal and generate simpler guards for these slow modules.
> 
> I've asked to create an issue about heuristics as well. Long-term, we want to start thinking about a low-level IR that can represent these optimization opportunities (or wether we can do this in the lowering itself.)

I'll actually implement this right now, it's quite a simple fix.

**Comment by rachitnigam** [comment]
Up to you! Let's not delay merging this PR too much if you're planning to travel this week.

**Comment by UnsignedByte** [comment]
Turned out to be a larger fix than I thought, oops!

File: src/ir_passes/lower/fsm.rs (hunk 2 above)
Comment (UnsignedByte) [reply src/ir_passes/lower/fsm.rs]: Done at #273.

Comment (rachitnigam) [reply src/ir_passes/lower/fsm.rs]: Sounds good!

File: src/ir_passes/lower/fsm.rs [hunk 8]
```diff
@@ -1,17 +1,540 @@
...
+/// Represents an fsm component.
+pub(super) struct FsmBind {
+    /// Fsm components using `n` counters each with `d` states, stored under `(n, d)`.
+    fsms: HashMap<FsmType, calyx::Component>,
+}
+
+impl From<(u64, u64)> for FsmType {
+    /// Generates an FsmType based on the default heuristic on the number of states and the delay (II).
+    fn from((states, delay): (u64, u64)) -> Self {
+        // TODO: Find a better metric to decide which type of fsm to generate.
```
Comment (rachitnigam) [inline src/ir_passes/lower/fsm.rs:33]: When adding `TODO` or `XXX`, please add your GH username: `TODO(rachitnigam): ...`

File: src/ir_passes/lower/fsm.rs [hunk 9]
```diff
@@ -1,17 +1,540 @@
...
+        self.fsms.entry(FsmType::Simple(states)).or_insert_with(|| {
+            let ports: Vec<calyx::PortDef<u64>> = (0..states)
+                // create the state ports in the format `_state`.
+                .map(|n| {
+                    (
+                        calyx::Id::from(format!("_{n}")),
+                        1,
+                        calyx::Direction::Output,
+                    )
+                        .into()
```
Comment (rachitnigam) [inline src/ir_passes/lower/fsm.rs:328]: PortDef constructions in Calyx is so ugly. We should probably create a macro for this in `calyx_ir`. Maybe worth opening an issue in the Calyx repo if you agree.

File: src/ir_passes/lower/fsm.rs [hunk 10]
```diff
@@ -1,17 +1,540 @@
...
+                ]);
+            }
+
+            drop(this);
+            comp
+        })
+    }
+
+    /// Generates a guard that is active for a range of states from start to end.
+    pub(self) fn range_guard(
```
Comment (rachitnigam) [inline src/ir_passes/lower/fsm.rs:430]: Perfect! This is exactly what I have imagining too!

**Review by rachitnigam — APPROVED** [review]
LGTM with at-discretion resolvable comments!

File: src/ir_passes/lower/fsm.rs (hunk 9 above)
Comment (UnsignedByte) [reply src/ir_passes/lower/fsm.rs:328]: Done! https://github.com/cucapra/calyx/issues/1668.

**Pull request merged**